pkg/core/native/native_neo.go

```diff
@@ -383,16 +383,28 @@ func (n *NEO) PostPersist(ic *interop.Context) error {
 	return nil
 }
 
-func (n *NEO) getGASPerVote(d dao.DAO, key []byte, index ...uint32) []big.Int {
-	var max = make([]uint32, len(index))
-	var reward = make([]big.Int, len(index))
-	d.Seek(n.ID, storage.SeekRange{Prefix: key}, func(k, v []byte) {
-		if len(k) == 4 {
+func (n *NEO) getGASPerVote(d dao.DAO, key []byte, indexes []uint32) []big.Int {
```

Comment:
I think you can make it
```
func (n *NEO) getGASPerVote(d dao.DAO, key []byte, index uint32) big.Int
```
now (avoiding slice allocations/sorting). We either need exactly one or two values from it and now we can position our Seek precisely, get one value and be done with it. Previously we had to go through the entire sequence of values, so doing just one Seek was important. Now making two Seek calls retrieving just one value is likely to be better (and much simpler), especially for big claim distances.

Reply:
I've implemented this optimisation in the last commit (as a separate commit), see the https://github.com/nspcc-dev/neo-go/pull/2316/commits/fd328b9916406b6bae1e017a60610645e3a0834e. But turns out that two separate `Seek` calls for `start` and `end` perform much worse than the old logic with a single `Seek` call for all the DBs, see the comparison with 4e76ce077db21e72286b2158b5cbac724687a89e.
```
    name                                             old time/op    new time/op    delta
    NEO_GetGASPerVote/MemPS_10_reward_records-8        29.3µs ± 3%    50.1µs ± 1%   +70.66%  (p=0.000 n=9+8)
    NEO_GetGASPerVote/MemPS_100_reward_records-8        221µs ± 9%     379µs ± 2%   +71.50%  (p=0.000 n=10+10)
    NEO_GetGASPerVote/MemPS_1000_reward_records-8      7.89ms ±93%   23.55ms ±62%  +198.57%  (p=0.000 n=10+10)
    NEO_GetGASPerVote/BoltPS_10_reward_records-8       26.3µs ± 2%    45.6µs ± 3%   +73.31%  (p=0.000 n=7+9)
    NEO_GetGASPerVote/BoltPS_100_reward_records-8      63.4µs ±15%   114.0µs ±19%   +79.89%  (p=0.000 n=10+10)
    NEO_GetGASPerVote/BoltPS_1000_reward_records-8      253µs ±13%     527µs ±18%  +108.38%  (p=0.000 n=10+10)
    NEO_GetGASPerVote/LevelPS_10_reward_records-8      29.5µs ± 9%    48.5µs ± 3%   +64.21%  (p=0.000 n=10+10)
    NEO_GetGASPerVote/LevelPS_100_reward_records-8     90.7µs ±16%   164.1µs ±17%   +80.97%  (p=0.000 n=10+10)
    NEO_GetGASPerVote/LevelPS_1000_reward_records-8     453µs ± 6%     735µs ± 8%   +62.31%  (p=0.000 n=10+9)
    
    name                                             old alloc/op   new alloc/op   delta
    NEO_GetGASPerVote/MemPS_10_reward_records-8        5.26kB ± 0%    9.11kB ± 0%   +73.15%  (p=0.000 n=10+9)
    NEO_GetGASPerVote/MemPS_100_reward_records-8       29.7kB ± 0%    57.7kB ± 0%   +94.15%  (p=0.000 n=10+10)
    NEO_GetGASPerVote/MemPS_1000_reward_records-8       280kB ± 8%     596kB ±16%  +112.83%  (p=0.000 n=10+10)
    NEO_GetGASPerVote/BoltPS_10_reward_records-8       5.53kB ± 1%    9.58kB ± 1%   +73.13%  (p=0.000 n=8+9)
    NEO_GetGASPerVote/BoltPS_100_reward_records-8      19.9kB ± 6%    37.8kB ± 8%   +89.51%  (p=0.000 n=10+10)
    NEO_GetGASPerVote/BoltPS_1000_reward_records-8      131kB ± 5%     262kB ± 5%  +100.84%  (p=0.000 n=10+10)
    NEO_GetGASPerVote/LevelPS_10_reward_records-8      6.28kB ± 2%   10.92kB ± 1%   +73.95%  (p=0.000 n=10+9)
    NEO_GetGASPerVote/LevelPS_100_reward_records-8     22.8kB ± 9%    44.4kB ± 9%   +94.53%  (p=0.000 n=10+10)
    NEO_GetGASPerVote/LevelPS_1000_reward_records-8     149kB ± 2%     290kB ± 3%   +94.57%  (p=0.000 n=10+10)
    
    name                                             old allocs/op  new allocs/op  delta
    NEO_GetGASPerVote/MemPS_10_reward_records-8           104 ± 1%       164 ± 1%   +57.67%  (p=0.000 n=10+10)
    NEO_GetGASPerVote/MemPS_100_reward_records-8          454 ± 3%       847 ± 4%   +86.35%  (p=0.000 n=10+10)
    NEO_GetGASPerVote/MemPS_1000_reward_records-8       5.03k ± 1%    10.04k ± 2%   +99.49%  (p=0.000 n=10+9)
    NEO_GetGASPerVote/BoltPS_10_reward_records-8          110 ± 1%       173 ± 0%   +57.70%  (p=0.000 n=10+8)
    NEO_GetGASPerVote/BoltPS_100_reward_records-8         484 ± 0%       910 ± 0%   +87.79%  (p=0.000 n=10+9)
    NEO_GetGASPerVote/BoltPS_1000_reward_records-8      4.15k ± 1%     8.23k ± 1%   +98.35%  (p=0.000 n=10+10)
    NEO_GetGASPerVote/LevelPS_10_reward_records-8         118 ± 3%       189 ± 2%   +59.90%  (p=0.000 n=10+9)
    NEO_GetGASPerVote/LevelPS_100_reward_records-8        488 ± 1%       910 ± 2%   +86.30%  (p=0.000 n=10+10)
    NEO_GetGASPerVote/LevelPS_1000_reward_records-8     4.21k ± 0%     8.39k ± 1%   +98.98%  (p=0.000 n=9+10)
```

I'm going to profile the last commit, but it seems to me that the `Seek` call itself is more resource-demanding than the code we have inside this call.

Reply:
Although claim distance for our benchmark is very small and equals to the committee size (6). Maybe we should try the benchmark with other claim distance?

Also note, that `Seek` still iterates over the whole set of values from the start, we don't perform early stop. We may refactor this code to use `SeekAsync`, but the backing store still has a lack of the early stop ability.

Reply:
Claim distance is important. 6 blocks is more of an exception for real networks than the rule. At the same time, for benchmark we have even shorter distance.

Reply:
We probably can stay with single Seek for now and move this question out to the new issue.